QUIC: server fixes for s_client compatibility

hlandau opened the pull request:
- Fix SSL_has_pending behaviour for compatibility with s_client. Since s_client uses SSL_has_pending to determine if I/O calls will be fruitful, returning 0 from SSL_has_pending can cause s_client to stall in server-writes-first scenarios, even if e.g. a FIN is waiting to be read.
- Fix server-side default stream creation.
- Ensure connections are not accidentally started before an application requests it.

Sashan commented on include/internal/quic_stream_map.h:
```diff
@@ -503,6 +503,40 @@ static ossl_inline ossl_unused int ossl_quic_stream_recv_get_final_size(const QU
+        if (!ossl_quic_rstream_available(s->rstream, &avail, &fin))
+            avail = 0;
+
+        if (avail == 0 && fin)
```
question/comment: I think it should actually count `fin` as byte here:
```
if (fin)
    avail++;
```

hlandau replied:
This only applies if avail is 0, so the outcome is the same either way...

Sashan replied:
The reason I was asking was we could potentially save one call to `ossl_quic_stream_recv_pending()` (I think). In case the fin as already received with last chunk of data (X+1).

It's a detail anyway.

Sashan commented on ssl/quic/quic_impl.c:
```diff
@@ -2577,6 +2580,10 @@ static int quic_write_nonblocking_epw(QCTX *ctx, const void *buf, size_t len,
     quic_post_write(xso, *written > 0, *written == len, flags,
                     qctx_should_autotick(ctx));
+    if (*written == 0)
+        /* SSL_write_ex returns 0 if it didn't read anything .*/
```
this looks odd to me. comment says `didn't read anything` but code belongs to `quic_write_nonblocking_epw()` function. Also is `SSL_ERROR_WANT_READ` at line 2585 the error code we want to return?

hlandau replied:
Good catch! Thanks.

Sashan reviewed (changes requested):
I've spotted just two places which I think need further clarification before approving. thanks a lot.

nhorman commented on include/internal/quic_stream_map.h:
```diff
@@ -503,6 +503,40 @@ static ossl_inline ossl_unused int ossl_quic_stream_recv_get_final_size(const QU
+    int fin = 0;
+
+    switch (s->recv_state) {
+    default:
```
you have all your ennumerated quic states specified as individual cases here.  Is it your intent to treat a value that is unknown (and falls into the default case), as a return 0?    Should that be an error?

hlandau commented:
Updated, changed my approach a little.

@mattcaswell I recommend you take a look at this PR. It is changing the behaviour of SSL_has_pending. In my view it is a bug fix - the existing behaviour of SSL_has_pending for QUIC didn't work right with the existing (non-QUIC-specific) s_client code. IMO s_client should be considered somewhat authoritative in terms of our API contract here.

The silver lining is that since SSL_has_pending returns a boolean and can return 1 due to e.g. yet-unprocessed packets, IMO erring on the side of returning 1 if there *might* be something to do seems safe to me and hard to see how it could cause breakage (there could always be something arriving that is not visible to the application causing spurious cases of SSL_has_pending returning 1, for both TLS and QUIC.)

Sashan reviewed (approved):
looks good to me.

t8m commented:
@hlandau should this go into 3.3 and 3.2 branches? IMO yes.

hlandau commented:
IMO yes.

hlandau commented:
This will not backport cleanly, created separate PRs:

https://github.com/openssl/openssl/pull/24039
https://github.com/openssl/openssl/pull/24040

t8m commented:
Merged to the feature/quic-server branch. Thank you.
